Cass wu5

richardkchapman opened the pull request.

richardkchapman committed:
HPCC-12251 Create cassandra plugin for workunit storage

Minor code cleanup.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-12251 Create cassandra plugin for workunit storage

Removed unused methods from workunitFactory interface.

Refactored any few remaining places that searched by xpath.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@jakesmith Please review

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2201,47 +2201,34 @@ void WsWuInfo::getWorkunitAssociatedXml(const char* name, const char* ipAddress,
         IConstWorkUnitInfo &cw = it->query();
         if (chooseWuAccessFlagsByOwnership(context.queryUserId(), cw, accessOwn, accessOthers) < SecAccess_Read)
             continue;
+        if (state && *state && !strieq(cw.queryStateDesc(), state))
```
Oddly, it post filtered on @state already, despite the xpath struff - see line 2239, so that is now a duplicate check.

richardkchapman replied:
Repushed to remove the duplicate

jakesmith commented on esp/services/ws_workunits/ws_workunitsService.cpp:
```diff
@@ -3112,8 +3049,10 @@ bool CWsWorkunitsEx::onWUExport(IEspContext &context, IEspWUExportRequest &req,
     try
     {
+        if (req.getECL() && *req.getECL())
+            throw makeStringException(0, "WUExport no longer supports filtering by ECL text");
```
seems a silly feature for WUExport anyway.

richardkchapman replied:
Yes - and was not used by our CLI client that called WUExport

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -2664,41 +2664,19 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
         return _getWorkUnitsByXPath(path.str(), secmgr, secuser);
     }
-    IConstWorkUnitIterator* getWorkUnitsByCluster(const char* cluster, ISecManager *secmgr, ISecUser *secuser)
+    IConstWorkUnitIterator* getScheduledWorkUnits(ISecManager *secmgr, ISecUser *secuser)
```
filtering by state is going to be a common one, even if done via generic mechanism now, I'd have thought.
Wouldn't keeping getWorkUnitsByState vs specific getScheduledWorkUnits be worthwhile during this slim down?

Although, I can imagine getScheduledWorkUnits might morph into an implementation which isn't scanning workunits at all..

richardkchapman replied:
In the dali case there's no difference. In the Cassandra case I'm going to have to be translating back to individual keys from the generic one to make sure that the common cases are efficient.

richardkchapman replied:
(and yes, I expect I may improve the implementation of getScheduledWorkUnits)

jakesmith commented:
Looks good, a couple of minor comments.
@richardkchapman

richardkchapman committed:
HPCC-12251 Create cassandra plugin for workunit storage

Remove duplicated code.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman committed:
HPCC-12251 Create cassandra plugin for workunit storage

Rationalize the lightweight workunit info, and its use in the WUQuery service.

Remove unused "RoxieCluster" value and references to it.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

jakesmith commented:
@richardkchapman - looks good/ready to merge.

richardkchapman commented:
I pushed an additional commit...

@ghalliday Please review

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented:
See also https://github.com/hpcc-systems/LN/pull/743

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -4236,11 +4179,10 @@ WUAction CLocalWorkUnit::getAction() const
     CriticalBlock block(crit);
-    str.set(p->queryProp("Action"));
-    return str;
+    return p->queryProp("Action");
```
Is it ok that this could be potentially invalidated by another thead?

ghalliday commented on common/workunit/workunit.cpp:
```diff
@@ -5765,7 +5707,7 @@ void CLocalWorkUnit::setStatistic(StatisticCreatorType creatorType, const char *
     }
 
-    if (mergeAction != StatsMergeAppend)
+    if (mergeAction != StatsMergeAppend) // RKC->GH Is this right??
```
I think it is correct, but part of the stats that doesn't feel quite right.  If appending the new values aren't combined with the existing ones (see code above which means that a new node is always created).

I think timestamps should possibly be revisited....

ghalliday commented on esp/services/ws_workunits/ws_workunitsService.cpp:
```diff
@@ -1875,18 +1808,49 @@ void doWUQueryWithSort(IEspContext &context, IEspWUQueryRequest & req, IEspWUQue
             continue;
         }
 
-        const char* wuid = cw.queryWuid();
```
@jakesmith what dos this do?  Does it remove dfu workunits from the list?

ghalliday replied:
Or is it something to do with the "global" workunit?

ghalliday commented:
@richardkchapman a few minor comments

richardkchapman committed:
HPCC-12251 Create cassandra plugin for workunit storage

Minor changes following review.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK
